stable_marriage/gale_shapley.nim

```diff
@@ -0,0 +1,274 @@
+      ["jon", "gav", "hal", "fred", "bob", "abe", "col", "ed", "dan", "ian"],
+      ["gav", "jon", "bob", "abe", "ian", "dan", "hal", "ed", "col", "fred"],
+      ["ian", "col", "hal", "gav", "fred", "bob", "abe", "ed", "jon", "dan"],
+      ["ed", "hal", "gav", "abe", "bob", "jon", "col", "ian", "fred", "dan"]
+    ]
+    # Recipient ids in descending order of preference
+    ContenderPrefs = initContenderPrefs(MPreferences, FNames)
+    # Preference score for each contender's id
+    RecipientPrefs = initRecipientPrefs(FPreferences, MNames)
+
+  proc randomPair(max: Positive): (Natural, Natural) =
+    ## Returns a random fair pair of non-equal numbers up to and including `max`
+    let a = rand(max)
+    var b = rand(max - 1)
+    if b == a: b = max
+    (a.Natural, b.Natural)
+
+  proc perturbPairs(ms: var Matches) =
+    randomize()
+    let (a, b) = randomPair(ms.contenderMatches.len - 1)
+    echo "Swapping matches between random contenders: ",
+          MNames[a], " <=> ", MNames[b]
+    template swap(arr: var openArray[int]; a, b: int) = swap(arr[a], arr[b])
+    ms.contenderMatches.swap(a, b)
+    ms.recipientMatches.swap(ms.contenderMatches[a], ms.contenderMatches[b])
+
+  func str(c: Clash; aNames, bNames: openArray[string]): string =
+    &"\u{0001F494} {aNames[c.id]} prefers {bNames[c.prefers]} over {bNames[c.match]}"
+
+  proc checkPairStability(matches: Matches; recipientPrefs, contenderPrefs:
+                                            openArray[Ranking]): bool =
+    let clashes = checkMatchingStability(matches, contenderPrefs, recipientPrefs)
+    if clashes.isSome():
+      let (clC, clR) = clashes.get()
+      echo "\u2717 Unstable\n", clC.str(MNames, FNames), '\n', clR.str(FNames, MNames)
+      false
+    else:
+      echo "\u2713 Stable"
+      true
+
+  proc render(contMatches: seq[int]; cNames, rNames: openArray[
+      string]): string =
+    for c, r in pairs(contMatches):
+      result.add(cNames[c] & " \u{0001F491} " & rNames[r])
+      if c < contMatches.high: result.add(", ")
+
+  var matches = stableMatching(ContenderPrefs, RecipientPrefs)
+  block:
+    echo render(matches.contenderMatches, MNames, FNames)
+    stdout.write "Current matching stability: "
+    let areStable = checkPairStability(matches, RecipientPrefs, ContenderPrefs)
+    doAssert areStable
+
+  perturbPairs(matches)
+  block:
+    echo render(matches.contenderMatches, MNames, FNames)
+    stdout.write "Current matching stability: "
+    let areStable = checkPairStability(matches, RecipientPrefs, ContenderPrefs)
+    doAssert areStable == false
+
+#===============================================================================
+{.push raises: [].}
+
+import std/options
+from std/sequtils import newSeqWith
+
+type
+  Ranking*[T: int] = concept c ## A wide concept allowing `stableMatching`
+                    ## to accept both arrays and sequences in an openArray.
+    c.len is Ordinal
+    c[int] is T
```

Comment:
Are `concept`s not experimental anymore?

Reply:
They are, but the old ones have been available and working for a very long time. I don't think there's anything wrong with using them, as I don't think there's a risk of them being dropped completely, only replaced by a new style concepts that we can adapt the code to as necessary.

Reply:
BTW, since this repo is educational we're much more free to use and showcase all the cool features of the language, compared to production situation.Since Nim doesn't have interfaces/traits, I think it's important to show the next best thing available.